### PR TITLE
Add sacct support and minor convenience functions

### PR DESCRIPTION
**Pull request by healther**
The first two commits are:
- minor refactoring of `status.py`
- option to write `output_specification` into a file
The latter two commits add:
- `SAcct`-Object which parses the output of `sacct` 
- adds printFinished functionality (using `SAcct`) and state and elapsed time information to the result

Depending on your wishes we could split this up into 3 PRs.

### CODE REVIEW

**Comment by terrycojones** [comment]
Hey, thanks for the pull request!  I just took a quick look through the code and it looks great - thanks also for following the Python style etc. I'll come back to this later today. Should be able to merge it quickly I think. Do you think there are tests that could be added, like the ones in `test/test_squeue.py`. It looks to me like that would make sense (I've never even used `sacct`, but I realize now that I was missing it the other day!). I am happy to add tests. Also we'll need to bump the version number in `__init__.py`. Thanks again :-)

**Comment by healther** [comment]
Currently we don't have any additional tests, feel free to add them. Not sure when/if I or someone from our group would find the time to add them.

When we are already talking about version numbers. You might want to take a look at your requirements.txt files. Are you really requiring the exact versions of your dependencies. I added a build description to [spack](https://github.com/spack/spack/pull/7681) and made the versions into open ranges.

**Comment by terrycojones** [comment]
@healther Are any of you SLURM experts?  I've spent quite a few hours working on your branch and I think it's about ready.

But looking at the output of `sacct --format JobID,Elapsed,State -P` I see lines like:

```
JobID|Elapsed|State
1174694|07:51:53|COMPLETED
1174694.batch|07:51:53|COMPLETED
1174694.extern|07:51:54|COMPLETED
1176831|12:00:22|TIMEOUT
1176831.batch|12:00:25|CANCELLED
1176831.extern|12:00:22|COMPLETED
1179221|00:00:27|FAILED
1180200|00:09:14|CANCELLED by 15053
1180200.batch|00:09:16|CANCELLED
1180200.extern|00:09:14|COMPLETED
1180200.0|00:09:15|CANCELLED
```

So the job id is often repeated and the dotted part (if any) is the job step. The code in the branch just overwrites the accumulating info for a job id (by dropping the part after the `.`, if any). It doesn't seem safe to do that. E.g., with job `1176831` in the above, the last line of its output shows it as `COMPLETED` but earlier lines show it as `TIMEOUT` and `CANCELLED`. So the job would end up looking like it had completed normally, when actually did not.

Also, do you need those methods such as `completed`, `failed`, `finished` for other purposes, or did you add them just because you could?

I'm tempted to look for `sacct` output lines that have no `.` in the job id and treat them as giving the most reliable summary of the job state. But maybe that's not correct. That also leaves the question of what to do if such a line isn't present but others are. The alternative is to collect the status info for all the job steps printed for a job id. That makes methods like `failed` a bit more complicated (they would check the status for each job step and if any were `FAILED` then return `True`). Hence my question about whether you need those functions. I presume yes, but what should the exact semantics be? I can probably come up with my own suggestion, but I'm wondering what you actually want/need.

Thanks!

**Comment by healther** [comment]
> Also, do you need those methods such as completed, failed, finished for other purposes, or did you add them just because you could?

I think they are mostly for documenting how to use `sacct` (convenience wrappers if you will, such that the user does not need to remember the sacct-conventions.

> I'm tempted to look for sacct output lines that have no . in the job id and treat them as giving the most reliable summary of the job state.

That seems to be reasonable (although I'm by no means a slurm expert). I'd probably be happy with having the option to turn stepwise information on/off. But I'm not the one who wrote this functions. CCing relevant people: @muffgaga @schmitts

**Comment by terrycojones** [comment]
@healther Thanks.

Thinking about this further, it seems to me I could get rid of all the `squeue` code and just use `sacct` for the status of all jobs. In some testing there was a very brief period where new job ids did not appear in `sacct` but that's no big deal.

**Comment by healther** [comment]
I think you are right, `sacct` should offer all information that `squeue` offers, up to some synchronisation issues, squeue should operate on the currently running stuff, which is more accessible than the global database. I'm not sure how this scales to large clusters with millions of jobs in their DB though.

**Comment by terrycojones** [comment]
@healther Thanks. I'm going to simplify the code and remove the `squeue` stuff (this will break backwards compatibility unless I leave in the `squeueArgs` option, which I doubt anyone is using, and which I don't even remember why I included - maybe for testing). I think the `sacct` database access should remain fast in almost all cases, due to the `-S` option. BTW, I modified your branch code so that by default it passes a value to `-S` that corresponds to the original submit time (which is in the JSON status) rather than just using a week. So it should all work well, I think. Thanks again for the pull request. I'll hopefully have all this done by the end of the weekend.

File: slurm_pipeline/sacct.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                            colsFound += 1
+                if colsFound != len(columnInfo):
+                    notFound = [title for title in columnInfo
+                                if columnInfo[title]['offset'] is None]
+                    raise SAcctError(
+                        "Required column%s (%s) not found in '%s' "
+                        'output' % ('' if len(notFound) == 1 else 's',
+                                    ', '.join(sorted(notFound)),
+                                    ' '.join(args)))
+            elif count == 1:
```
Comment (terrycojones) [inline slurm_pipeline/sacct.py]: What version of SLURM do you need this for? In the version I have access to (17.02.10) there is only one header line.

Comment (healther) [reply slurm_pipeline/sacct.py]:
```
$ sinfo -o "%v"
VERSION
17.02
```

But when I simply do 
```
$ sacct
       JobID    JobName  Partition    Account  AllocCPUS      State ExitCode 
------------ ---------- ---------- ---------- ---------- ---------- --------
```

If I do
$ sacct -p -l -u sschmitt
JobID|JobIDRaw|JobName|Partition|MaxVMSize|MaxVMSizeNode|MaxVMSizeTask|AveVMSize|MaxRSS|MaxRSSNode|MaxRSSTask|AveRSS|MaxPages|MaxPagesNode|MaxPagesTask|AvePages|MinCPU|MinCPUNode|MinCPUTask|AveCPU|NTasks|AllocCPUS|Elapsed|State|ExitCode|AveCPUFreq|ReqCPUFreqMin|ReqCPUFreqMax|ReqCPUFreqGov|ReqMem|ConsumedEnergy|MaxDiskRead|MaxDiskReadNode|MaxDiskReadTask|AveDiskRead|MaxDiskWrite|MaxDiskWriteNode|MaxDiskWriteTask|AveDiskWrite|AllocGRES|ReqGRES|ReqTRES|AllocTRES|
1116681|1116681|bash|short|310860K|HBPHost6|0|29292K|3260K|HBPHost6|0|2224K|3K|HBPHost6|0|3K|00:00:00|HBPHost6|0|00:00:00|1|2|00:01:31|COMPLETED|0:0|1.10G|Unknown|Unknown|Unknown|3750Mc|0|0.84M|HBPHost6|0|0.84M|0.01M|HBPHost6|0|0.01M|||cpu=1,mem=3750M,node=1|cpu=2,mem=7500M,node=1|
```
I also get no separation line... It's probably better to do the checking explicitly

Comment (terrycojones) [reply slurm_pipeline/sacct.py]: Huh, that's a bit weird. I'll do the checking, as you suggest. Thanks.

**Comment by terrycojones** [comment]
In case you're curious, I have a branch at https://github.com/acorg/slurm-pipeline/tree/electronicvisions-PR/upstream with the switch to using `sacct`. I haven't actually used it yet, though there are tests for everything.  I'm going to play with it tomorrow and if everything seems ok I'll merge it and update on PyPI.  I've upped the version number to 2.0.0 because the change is not backwards compatible (due to dropping the `--squeueArgs` option).

Anyway, if you want to take a look, feel free. I'll of course be interested to hear if it provides everything you were aiming for in your branch.

Thanks again!

**Comment by terrycojones** [comment]
I've made a new pull request, based on this one, at https://github.com/acorg/slurm-pipeline/pull/30

I plan to merge it & release to PyPI later today. I've spent some time using it today and so far it seems to work. The status command does feel slow, though. I'll have to check that that's due to `sacct --jobs` and if so we can decide if we can live with it.

I plan to close this pull request.

**Comment by terrycojones** [comment]
I merged my branch. Closing this. Thanks again.

**Comment by healther** [comment]
Thanks I'll let our people know they should try with your new version and will report back any problems that we will find

**Pull request merged**